### PR TITLE
webapp changes for custom software images

### PR DESCRIPTION
**Pull request by haraldschilly**
# Description
This will contain front-end modifications for custom software environments. So far

- show the name of the chosen image

# Testing Steps
1.
1.
1.

# Relevant Issues

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):
- [ ] No debugging console.log messages.
- [ ] All new code is actually used.
- [ ] Non-obvious code has some sort of comments.

Front end:
- [ ] Restart at least one project and check its `~/.smc/local_hub/local_hub.log`
- [ ] Completely restart Webpack with `./w` in `/src`
- [ ] Completely restart the hub by killing and restarting `./start_hub.py` in `/src/dev/project`
- [x] Screenshots if relevant.

### in project settings, explaining which features are available

**default**
Screenshot from 2019-04-13 12-24-39

**old image, from a year ago**
Screenshot from 2019-04-13 12-26-20

**custom software, in that case it is just for a specific python library**
Screenshot from 2019-04-13 12-23-35

### CODE REVIEW

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 5212dd76527f21fd8c4977470ff70e8e801c1ef9 into b5505f928439dac38bf4ec57e224807230501c21 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-59bba0dd91f456df80275b5999387945687d8888)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 4 alerts** when merging 9c0421a266f42070f59d746aabb9bc25a1c0e648 into 869bf7079c621215632daeaf0a539ce73c2ddef5 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-ae2a4ee8a37e921e6cc88f062d827ad1da8f7cbe)

**new alerts:**

* 2 for Unused variable, import, function or class
* 1 for Potentially unsafe external link
* 1 for Invocation of non\-function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 4 alerts** when merging 77f926528022e70f29e105975d06e9b2d44a9607 into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-79cb44fe24345ee4389f8fd6c0d1a72cc6296e02)

**new alerts:**

* 2 for Unused variable, import, function or class
* 1 for Potentially unsafe external link
* 1 for Invocation of non\-function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 5488dfa914be717857a29214cdd49bf4399550df into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-a53b436e2fc815275fdf3007f8ccc5f2402a058e)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging b45ee0b6cbcd98f4457de961a4e4eb633a931b9f into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-1e27d3f9d5afac22229801289ff0cdb64c3e218e)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 3b81171f414ff2b30598754feccae3a2a90834a3 into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-5d5fe4e7f4f27ca602fc717805ec733a40e2d1e3)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging fb3b8906c62dd12dfd73b581b17b536438352004 into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-ef93e9089832cdf6c2b421f0ca5e8ac2d83f5965)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging bdb153480d09d300ea53eb18df0d797c15525437 into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-8a49dbc3ceafa2e87a710794bf42d481b1ad58fd)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 1880df73372ca35c24a3f18edf4741a692897e8c into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-4dacd173c4f1c148949c835334530568a8d48cbb)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging c5e10ab30ee340b60a63e2739477e2143a127bce into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-4f7e79cb1d01fa65bcdc700ed8b18da32ac4d4f1)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging a54743a1eff786569c185e034fc05d40028a2c24 into 5eeb5243c694840d36c97227b5eb100e8a1fb668 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-498a1cc38e077082af77cc43f9c7667163fbf63f)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging b4da2616db405c3d67101e819776417b3358b7db into 0a3e65531257d06360217fa4df1c78fd83cbf543 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-29168ce52740577a845750f07d40763b443119a7)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging f26c1c598a9519d8ab039d14c8ef5f5b1f781ef1 into 0a3e65531257d06360217fa4df1c78fd83cbf543 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-5e854a87ddb60f0f966adfdb3f5afa6947acc3c0)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging cba922ff05b82fa02eae267d21dd757028ba547f into 8b52b659288f1c45096004750671b7f15932f503 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-103e3f941d42e8f04712475a37dbc703842f27f6)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging ffe26f24d9f8bd063727295e51ef290b10f773ff into 8b52b659288f1c45096004750671b7f15932f503 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-006df6de653a677a3892399f7ceca15c2a7dcce7)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 08e01974ab9dbc8cbb48f3101116fff9af2e94fd into 8b52b659288f1c45096004750671b7f15932f503 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-4f7134387590831c224643a6a52a7cbb921db24d)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 2 alerts** when merging ecd86e1c358942045827cc0bd5a9273366b6163a into 4444232e3251b16ccee70a96e4d506e055487451 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-1c15cfc59b0287279ce777ce62b9fe66a307e507)

**new alerts:**

* 1 for Conflicting HTML element attributes
* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging fbca6387869b0df643787fff59350fe92f57769b into 4444232e3251b16ccee70a96e4d506e055487451 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-ea5b55a9972075f52cf33ccfe17a811df08c4f66)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 54ce864e95771b37487485860fecf253083e2ab0 into 4444232e3251b16ccee70a96e4d506e055487451 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-9bb140cce8ecf6333aa06a544de78390a292cdcd)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 517fccfcbda0efa6459d27853cb3b06681e95473 into 4444232e3251b16ccee70a96e4d506e055487451 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-4904dfab261605086d0cf7e09c601f1f6a3639c2)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 5b96b9b730814b51111b18c072ed7b7cc9eae93a into 035e22b92bad320e904ee5d990bc8a6a365cf2e8 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-65e5f2a6dbaa9902c0c44b92526717ed7869ac8b)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging f783260795f621dade1ab54f760da338340db5c8 into 035e22b92bad320e904ee5d990bc8a6a365cf2e8 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-4765fa2cdd5f7ce7624231c25104cf7ddf596ed3)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 6ad1d026a2cb2b91f3c92d320a7f4fc0a98ad5bc into 58d650301fc26b563c5ef7261c2786dc67378260 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-6387aaa1d6b055d731af5250aea92dcacfc99947)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 82443e3857617f5f07b01540177afaa7df4521af into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-39f5370d5ab394d52ac6aa3a6041b60bc59e9aff)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging e84f1ce0cafebe0b6321be12b11d750dfdcc12bf into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-9c2231c5a8d0e9356a6b7230dc14ca3ed6fe4164)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 2 alerts** when merging 133b32fdb912643f042ef8f47073c7575c922803 into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-cda6b959c6a9a60bdcd0333f19eb14ea73901724)

**new alerts:**

* 1 for Overwritten property
* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 57cc5bd1385666aabae0729593dd11212d2b91e1 into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-1edc0826b294e9be370107ac3d7c53d87dd02693)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 2e3aa767a9078d919e9610087f4bc504db7f7d68 into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-96b7a0cde5ba823143759cd332d33ebc8825c087)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging dfde96af3937c1beae12713fdd7e4ba27ad091fa into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-08e119d34d09cfc325f5bc007a7dc786bf9902fa)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging dc7b32c0371a3873f7d55f2a0ce5aaacdfe936a3 into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-64693832b68c6467240b069cff134e0ffebceea4)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging f2fb2b48c3cb2c1e85f4fdaec24842e3f5765d25 into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-3aea5662652dce89ea34e69c1b4ddccf6a62ac1f)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 8270bfe43aac714321b565be9412b708400d5cd6 into 4162ae3c56d6cb2fa67e85f413fb131892104ea7 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-86fc17db54107dc0e2bb2bf95fdc1cb60af2fc01)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 7ee3a0ce410c18e2ed21ab092db23c65462a1be8 into 7bb5050da3965da048de682f2461add67ef52ea5 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-74d3d7bf2163bfe8dd3777da90642234f0f4de95)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 2547e51bd68f540f1076d3730ba7a3bae7b22bd9 into 7bb5050da3965da048de682f2461add67ef52ea5 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-847f929e6664d9e95fb767de7bcc018260002a1b)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 76d259f63dc38ec6e3824953dce55af2659db209 into 9f47e008c3503849a31490e9bbf790c04d829211 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-f01e02d9c7c2fa38f05e0f1d3f638539cab2d53e)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 8e7fae2250bef4e6dddb02e097118e6fbee80e8c into 9f47e008c3503849a31490e9bbf790c04d829211 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-afea37b29bbbfc7911ea749fa6697637cea6d1a0)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging be61f5f7f76c3b8c46ef48e25488fa353244c6c6 into 9f47e008c3503849a31490e9bbf790c04d829211 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-5429382ca68c18c9cc7dc64a62c1531929162e42)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 6a9c1aefe7c3eea407c5a055ce231c7f2428ef5d into 9f47e008c3503849a31490e9bbf790c04d829211 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-4c9f563ca93100cbaf5105863e8e996762b30a90)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 10724ee704636abcda8e8d5e440c80aa2f948128 into 20c68473e5b6605984e669ff117751aa9be1e2a1 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-29c9c3beffc5cd32d044b423cf9f8aab8f84e93f)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 4f5dee5dea3fec18be2b50dfbc2810d7db388115 into 20c68473e5b6605984e669ff117751aa9be1e2a1 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-6062c1ef8fbb8a475dee5fb9ac4e1f69aca7d981)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 90a8aeea16dfbe83b0a6595dab78eca4f6cacea1 into 20c68473e5b6605984e669ff117751aa9be1e2a1 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-68a36357a5af6e102c115dcd3ade9065df37b225)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 2 alerts** when merging 08c525fb2ee025c21abec3a3ff33b180d3bc2315 into 20c68473e5b6605984e669ff117751aa9be1e2a1 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-f0097c10b85e15fedc56d2054b0d48111362f2f9)

**new alerts:**

* 1 for Potentially unsafe external link
* 1 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 679be006227d0e404f8e397c856cc82517686d69 into 20c68473e5b6605984e669ff117751aa9be1e2a1 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-8993555c0e6adde47357b2a66f7c53a72dc3770d)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 73065478cd3971fc0f3e6c70f327661a9e23fd6d into d82fc0a4e41f2bc33323a2a5d3b7a8af936ef8d3 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-7ef29c58e87b500a06f8e8faef10ec4244fa139a)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 34f0635c18d436fb4ab50d178b45e3d84d8fa591 into d82fc0a4e41f2bc33323a2a5d3b7a8af936ef8d3 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-99cd6f859644ae8ea7cfb5f32da36f19cbba3916)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 10ca66b816b8a12ae5d0c56e358a1c858ac6ccc5 into d82fc0a4e41f2bc33323a2a5d3b7a8af936ef8d3 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-02690b9769d2d94780d36314c33bfec05ea7cf1e)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 438cb0b8c0e51aa3025491a740c148b1b5812875 into d82fc0a4e41f2bc33323a2a5d3b7a8af936ef8d3 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-8bdba30b9c51605be996167c8ecaa53429510980)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 87d633186ccdd9f031ccc1ec1880bdd117a5fefe into 02bfc1e7041ca2d9f19950c10ad37a6d3003b1f5 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-43211458c45441f840384e0768a7a3007d836431)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 8fb29170443a7ff692bb6e8002c91103856adfe8 into eef16d2607b687b4d8a91ad2deb78313a6c1ecbc - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-9eba48b112217aef66424ad1a0fc50f7f8c0d971)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 2c2c1909df15581c4ba67c07739255bd55e5db4c into d989b5823b0f2ace0f04c4d912505e23cf719b02 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-d3c7841b66f0410373cd3acc2fea967cda63f40b)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging 781f8457e146f1eee73f0bbce4a2a9a9bff8e7a9 into eb6b505cfb0202f71d0cbe1823cdb8ae255d9353 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-d54da7e58e9399a23065662e152ec0e3579aac12)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by haraldschilly** [comment]
This pull request **introduces 1 alert** when merging c8a49de48dba6b28f313a0b3855cc24b7e8b646f into a4304fe05029fe03e0a32de1bf7d810f3190d785 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-023fd7f07a493ad2476d93ebbb30ed580f4e7415)

**new alerts:**

* 1 for Potentially unsafe external link

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/smc-webapp/project_configuration.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
+/*
+ * this manages project configuration specific aspects.
+ * It is the corresponding counterpart of smc-project/configuration.ts
+ * The various "capabilities" data-structures are used to show/hide UI elements or suppress
+ * calling certain operations which are not possible (e.g. spellcheck requires aspell)
+ */
+
+import { Map as iMap } from "immutable";
```
Comment (johnjeng) [inline src/smc-webapp/project_configuration.ts]: Still don't like this.

File: src/smc-webapp/project_configuration.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    if (!caps.latex) disabled_ext.push(...KNITR_EXTS.concat(["tex"]));
+    if (!caps.sage) disabled_ext.push("sagews", "sage");
+    if (!caps.x11) disabled_ext.push("x11");
+  }
+
+  if (prev != null) {
+    const next = prev.set(aspect, config);
+    return next;
+  } else {
+    return iMap<ConfigurationAspect, Configuration>([[aspect, config]]);
```
Comment (johnjeng) [inline src/smc-webapp/project_configuration.ts]: And at this point iMap is just wrong...

File: src/smc-webapp/custom-software/selector.tsx [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    if (this.props.title_prefill) {
+      // keep title_prefill as it is
+      this.props.setParentState({ title_text: display });
+    }
+  };
+
+  render_custom_image_entries() {
+    if (this.props.images == null) return;
+
+    const search_hit = (() => {
```
Comment (johnjeng) [inline src/smc-webapp/custom-software/selector.tsx]: Why is this an IIFE?

Comment (johnjeng) [reply src/smc-webapp/custom-software/selector.tsx]: Just to make `search_hit` a const?

File: src/smc-webapp/custom-software/init.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  | "type"
+  | "display"
+  | "url"
+  | "desc"
+  | "path"
+  | "search_str"
+  | "display_tag"
+  | "disabled";
+
+export type ComputeImage = iMap<ComputeImageKeys, string>;
```
Comment (johnjeng) [inline src/smc-webapp/custom-software/init.ts]: It doesn't seem to make tooo much a difference here but we have a type for this usage:

```
import { TypedMap } from "../app-framework/TypedMap";

ComputeImage = TypedMap<{
  id: string,
  src: string,
  // ... etc
}
```

Then you can specify some of them as optional. As written in your code now, they're all optional.
Also allows you to do stuff like 
```
keyof ComputeImage
```

File: src/smc-webapp/project_configuration.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  latex: true,
+  rmd: true,
+  x11: true,
+  spellcheck: true,
+  library: true,
+  formatting: true
+});
+
+// detecting certain datastructures, only used for TS typing
+function isMainCapabilities(
```
Comment (johnjeng) [inline src/smc-webapp/project_configuration.ts]: Nice using a custom type checker.

File: src/smc-webapp/custom-software/selector.tsx [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+interface CSState {
+  search_img: string;
+}
+
+const CS_INIT_STATE: Readonly<CSState> = Object.freeze({
+  search_img: ""
+});
+
+const cs_list_style: Readonly<React.CSSProperties> = Object.freeze({
```
Comment (johnjeng) [inline src/smc-webapp/custom-software/selector.tsx]: Freezing CSS properties seems kind of silly to me. What's the worry here? Or do you just wish (as I do) that javascript had better immutable objects?

File: src/smc-webapp/project_files.cjsx [hunk 7]
```diff
@@ -45,7 +47,26 @@ underscore            = require('underscore')
...
+# treat this as const/readonly.
+# the order of these buttons also determines the precedence of suggested file extensions
+# see project/file-listing/utils.ts
+ALL_FILE_BUTTON_TYPES = exports.ALL_FILE_BUTTON_TYPES =
+    Object.freeze(['sagews', 'ipynb',  'tex', 'term',  'x11',
+        'rnw', 'rtex', 'rmd', 'md', 'tasks', 'course', 'sage', 'py', 'sage-chat'
+    ])
+
+
+ROW_INFO_STYLE = Object.freeze
```
Comment (johnjeng) [inline src/smc-webapp/project_files.cjsx]: This just doesn't even seem necessary since it's not exported.

**Pull request merged**